README.md
# README #
### Copyright ###
Poppins 

Copyright (C) 2015 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by Bruno Dooms, Frank Van Damme

### Summary ###
* Quick summary: Poppins - backup script with incremental snapshots. 
* Version: Poppins version 0.2 beta

### Install ###
Step 1. Make sure following packages are installed on the backup server: 

    mercurial, php5-cli (php-cli), rsync, ssh, grep, gzip, moreutils

Step 2. Verify if the timezone is configured correctly in php. Look for a php.ini file in the /etc directory. For example:

    date.timezone = Europe/Brussels

Step 3. Download the source code with the hg command. 

    hg clone https://bitbucket.org/poppins/poppins /opt/poppins

Step 4. Make a link to init.php in /usr/local/bin.

    ln -s /opt/poppins/init.php /usr/local/bin/poppins

Step 5. Verify the installation.  

    poppins -v

Step 6. Make sure following packages are installed on remote machine: 

    rsync, ssh, grep, aptitude (Debian) or yum-utils/rpm (Red Hat). 

Step 7. Establish a passwordless ssh login to the client using ssh-keygen & ssh-copy-id.

### mysql ###
If using mysql backups, credentials must be provided in the .my.cnf file. See config.

### Upgrade ###
Navigate to the poppins source directory and pull the code with the hg command: 

    hg pull -u


### Configuration ###
All configuration in {filename}.poppins.ini. See example.poppins.ini for instructions.

### Deployment ###
SYNOPSIS

    poppins -c {configfile} [--color] [-d] [-h] [-n] [-t {tag}] [-v] [--long-options]

DESCRIPTION

Poppins has one required option, "-c", the configfile. An example can be found in the poppins source directory. You must name it {filename}.poppins.ini. Options in the ini file may be overridden by cli options. See example.

EXAMPLES

    poppins -c example.poppins.ini --remote-host=webserver1 --color

OPTIONS

    -c {configfile}
        Required configuration file.

     --color
         Output in pretty colors.

    -d
        Debugmode. Output all commands ran by Poppins.

    -h, --help
        Print this help page.

    -t {tag}
        Add a tag to the application log file.
        
    -n
        Dry run

    -v
        Print Poppins version.

### Cleaning up old log files ###

SYNOPSIS

    scripts/logdir.cleanup.sh  -l {logdir} [ -a ]

DESCRIPTION

This script serves the purpose of cleaning up log old log files. Once old snapshots are rotated away, the log files from the backup jobs that created them are no longer useful. logdir.cleanup.sh will analyze the log files of succeeded backups, and propose to delete those whose snapshots have disappeared from your system. Optionally, it will also let you remove the logs of failed backups.

Take care when removing logs from backups that ended with an ERROR state! They might contain useful information about how or why your backup failed. Since a backup job in ERROR state never has snapshots (Poppins does not create snapshots based on a failed backup), logdir.cleanup.sh has no way of telling which log files are old enough to delete!

EXAMPLES

    logdir.cleanup.sh -l /root/poppins.d/logs 
    logdir.cleanup.sh -l /var/log/poppins -a

OPTIONS

    -l {logdir}
         Required: directory where you configured Poppins to save its log files.

    -a
        Automatic: old log files will be deleted, but logs of failed backup jobs will be left alone.



release_notes.md
# RELEASE 0.4

DATE: 2019-07-19

## NEW FEATURES

### INI FILE

The following ini configuration directives need to be ADDED to the config file.

See example.poppins.ini:

    ...
    [remote]
    ; Ssh remote port, default is 22
    port = 22
    ...
    [meta]
    ...
    ; include restore scripts
    restore-scripts = yes
    ...
    [rsync]
    ...
    ; add timestamps to every line of rsync output
    timestamps = yes
    ...
    [mysql]
    ...
    ; include specific databases
    ; not supported in combination with multiple .my.cnfs
    ; you may use a regular expressions (Perl synxtax) within slashes "/"
    ; e.g. '/^drupal_.*/' # all databases starting with string 'drupal_'
    ; default is every database except information_schema
    ; included-databases = 'wordpress_1,wordpress_2,/^drupal_.*$/'
    
    ; exclude databases - see previously
    ; excluded-databases = '/^drupal_test.*/'
    
    ; ignore specific tables within the databases
    ; Use dot notation "database.table" or regex "/<REGEX>/" (Perl synxtax)
    ; e.g. '/database1\.tbl.*/' # all tables in database1 starting with string tbl
    ; default is every table except mysql.event
    ; ignore-tables = '/^wordpress1\.test_.+/'
    
    ; include specific tables in "tables" or "csv" output
    ; not supported in combination with multiple .my.cnfs
    ; Use dot notation "database.table" or regex "/<REGEX>/" (Perl synxtax)
    ; e.g. '/^database1\.tbl.*$/' # all tables in database1 starting with string tbl
    ; included-tables = '/^wordpress4\.tbl_.+/'
    
    ; exclude tables - see previously
    ; excluded-tables = ''
    
    ; output types: databases dumps, table dumps or tab seperated file (csv)
    ; use values database, table or csv (or combination)
    output = 'database,table,csv'

## WHAT HAS CHANGED?

### MAJOR CHANGES

* Include or exclude specific databases or tables, based on regular expressions.
* MySQL export options: database dump, separate table dumps and/or csv.
* Added restore scripts to help with restoring backups.
* Keep snapshots with underscore indefinitely. Ignore these snapshots from rotation.

### MINOR CHANGES

* Type support/validation enabled in config file
* Rsync error 24 triggers a notice, no longer a warning.
* Show version hash when calling Poppins version.
* Run remote ssh session on a different port.

## WHAT HAS BEEN FIXED?

* Small bugfixes and cosmetic changes

## KNOWN ISSUES

## INSTALLATION

To upgrade, run following command in the poppins directory:

    hg pull -u

------------------------------------------------------------------------------------------------------------------------
# RELEASE 0.3

DATE: 2017-06-30

## NEW FEATURES

### INI FILE

The following ini configuration directives need to be ADDED to the config file.

See example.poppins.ini:

    [remote]
    ...
    ; if the backup job fails, abort or continue with the post-backup script
    ; values are "abort" or "continue"
    backup-onfail = "abort";
    ; remote script ran after backup
    ; e.g. post-backup-script = "/home/poppins/some-post-backup-script.sh";
    post-backup-script = "";
    ...
    [rsync]
    ...
    ; cross filesystem boundaries
    cross-filesystem-boundaries = no

## WHAT HAS CHANGED?

### MAJOR CHANGES

* Directive "cross-filesystem-boundaries" is added to the rsync section. A warning is triggered when trying to cross mounted filesystem boundaries if this option is set to "no", which is the default. You must explicitly exclude these directories or set the option to "yes".
* Notices are introduced. Less important messages (dir not clean, incomplete configuration, duplicate mysql config file) will be considered notices rather than warnings.
* Post backup script added. May or not be run depending on a successful rsync run.

### MINOR CHANGES

* Cleanup script added to remove old log files. See scripts directory.
* Validation of trailing slashes in exluded/included sections for consistency.
* Validation of relative paths in excluded section as the exclude path is always treated as a relative path by rsync.

## WHAT HAS BEEN FIXED?

* Small bugfixes and cosmetic changes

## KNOWN ISSUES

* Validation of ini files is imperfect because of the lack of type support of
the function parse_ini_file(). As of PHP 5.6, INI_SCANNER_TYPED needs to be
implemented when PHP 5.6+ is available on Debian and CentOS latest releases.

## INSTALLATION

To upgrade, run following command in the poppins directory:

    hg pull -u

------------------------------------------------------------------------------------------------------------------------
# RELEASE 0.2

DATE: 2016-07-01

## NEW FEATURES

### INI FILE

The following ini configuration directives need to be ADDED to the config file.

See example.poppins.ini:

    ...
    [local]
    ; snapshots are created using hardlinks by default
    ; if available, you may use zfs or btrfs
    ; filesystem snapshots instead
    ; values are "default", "btrfs" or "zfs"
    snapshot-backend = "btrfs"
    ...
    [remote]
    ; enable remote connection with ssh, default is yes
    ; disable if backing up local directories (e.g. to an external drive)
    ; host and user directives are disregarded in case ssh is disabled
    ssh = yes
    ; ssh connection retries if desired
    ; check is applied on first connection attempt
    ; default is 0
    retry-count = 0
    ; timeout between retries in seconds
    ; default is 0
    retry-timeout = 10
    ...

The following directives need to be REMOVED:

    ...
    ; local filesystem options: default/ZFS/BTRFS
    ; use ZFS or BTRFS if you want to use shapshot features
    ; for these filesystems. Otherwise, use default.
    filesystem = 'BTRFS'
    ...

### NEW CLI OPTIONS

The following cli options have been added:

* --color: add colors to cli output
* -t {tag}: add an optional tag to poppins log file

### TAGGING

You may add a tag to a poppins run, e.g. a hash or timestamp:


dates.php
<?php
function to_time($stamp, $format = 'unix')
{
    $t = explode('_', $stamp);
    $date = $t[0];
    $time = implode(':', str_split($t[1], 2));
    $datetime = $date . ' ' . $time;
    switch($format)
    {
        case 'unix':
            $result = strtotime($datetime);
            break;
        case 'string':
            $result = date($datetime);
            break;
    }
    return $result;
}
//check if a diff exceeds a himan readable value
function time_exceed($diff, $type)
{
    global $DIRS;
    if(!in_array($type, $DIRS))
    {
        die('Period not supported!');
    }
    if(!is_integer($diff))
    {
        die('Diff must be an integer!');
    }
    $seconds['hourly'] = 60*60;
    $seconds['daily'] = $seconds['hourly']*24;
    $seconds['weekly'] = $seconds['daily']*7;
    $seconds['monthly'] = $seconds['daily']*30;
    $seconds['yearly'] = $seconds['daily']*365;
    return (boolean) ($diff >= $seconds[$type]);
}

$DIRS = ['incremental', 'hourly', 'daily', 'weekly', 'monthly', 'yearly'];
//initiate arrays
foreach($DIRS as $c)
{
    $arch1[$c] = [];
    $arch2[$c] = [];
}

$arch1 ['incremental'] []= 'jessie2.2015-06-05_170100.poppins';
$arch1 ['incremental'] []= 'jessie2.2015-06-06_170100.poppins';
$arch1 ['incremental'] []= 'jessie2.2015-06-07_170100.poppins';
$arch1 ['incremental'] []= 'jessie2.2015-06-08_170100.poppins';
$arch1 ['incremental'] []= 'jessie2.2015-06-09_170100.poppins';

$arch1 ['hourly'] []= 'jessie2.2015-06-10_170100.poppins';
$arch1 ['hourly'] []= 'jessie2.2015-06-10_180100.poppins';
$arch1 ['hourly'] []= 'jessie2.2015-06-10_190100.poppins';

//$arch1 ['daily'] []= 'jessie2.2015-06-07_170100.poppins';
//$arch1 ['daily'] []= 'jessie2.2015-06-08_170100.poppins';
//$arch1 ['daily'] []= 'jessie2.2015-06-09_170100.poppins';
//
//$arch1 ['weekly'] []= 'jessie2.2015-06-09_170100.poppins';

$arch1 ['monthly'] []= 'jessie2.2015-05-31_170100.poppins';
$arch1 ['monthly'] []= 'jessie2.2015-06-01_170100.poppins';

//$arch1 ['yearly'] = [];
//$arch1 ['yearly'] []= 'jessie2.2015-06-09_170100.poppins';

//copy archive 1
$arch2 = $arch1;

$settings['snapshots']['incremental'] = 0;
$settings['snapshots']['hourly'] = 2;
$settings['snapshots']['daily'] = 1;
$settings['snapshots']['weekly'] = 1;
$settings['snapshots']['monthly'] = 2;
$settings['snapshots']['yearly'] = 1;

$cdate = '2017-06-10_200200';

//echo variable
$newdir = 'jessie2.' . $cdate . '.poppins';
echo "CURRENT DATE IS $cdate\n";

#####################################
# SORTEREN
#####################################
//sort alphabtically
echo "\n-----------ORIGINAL----------------\n";
foreach ($arch2 as $k => $v)
{
    echo "\n[$k]\n";
    if(count($arch2[$k]))
    {
        asort($arch2[$k]);
        $base[$k] = $arch2[$k][0];
        $end[$k] = (end($arch2[$k]));
    }
    else
    {
        $base[$k] = '';
        $end[$k] = '';
    }
    //
    foreach ($arch2[$k] as $vv)
    {
        echo "$vv, ";
    }    
}
#####################################
# AFLOPEN
#####################################
foreach ($arch2 as $k => $v)
{
    if ($k == 'incremental')
    {
        $arch2[$k] [] = $newdir . '.NEW';
        if($settings['snapshots'][$k]) $move[$k] = true;
    }
    else
    {
        if ($end[$k])
        {
            if (!preg_match('/[0-9]{4}-[0-9]{2}-[0-9]{2}_[0-9]{6}/', $end[$k], $m))
            {
                die("WRONG FORMAT, CANNOT CONTINUE!");
            }
            else
            {
                $ddate = $m[0];
                $cdate2unix = to_time($cdate);
                $ddate2unix = to_time($ddate);
                if($cdate2unix < $ddate2unix)
                {
                    die('Cannot continue. Newer dir found: '.$ddate);
                }
                else
                {
                    $diff = $cdate2unix - $ddate2unix;
                    if (time_exceed($diff, $k))
                    {
                        $arch2[$k] [] = $newdir . '.NEW';
                        if($settings['snapshots'][$k])  $move[$k] = true;
                    }
                }
            }
        }
        else
        {
            $arch2[$k] [] = $newdir . '.NEW';
            if($settings['snapshots'][$k]) $move[$k] = true;
        }
    }
}

#####################################
# SLICE
#####################################
echo "-----------RESULT----------------\n";
foreach($arch2 as $k => $v)
{
    $n = $settings['snapshots'][$k];
    $arch2[$k] = array_slice($arch2[$k], -$n, $n);
    echo "\n[$k]\n";
    foreach($arch2[$k] as $vv) echo "$vv, ";
}
#####################################
# COMPARE ARCH TO RES
#####################################
$add = [];
$remove = [];

foreach($arch2 as $k => $v)
{
    if(count($v))
    {
        foreach ($v as $vv)
        {
            $remove[$k] = array_diff($arch1[$k], $arch2[$k]);
            $add[$k] = array_diff($arch2[$k], $arch1[$k]);
        }
    }
    else
    {
        $remove[$k] = $arch1[$k];
    }
}
    
echo "\n-----------REMOVE----------------\n";
if (count($remove))
{
    foreach ($remove as $k => $v)
    {
        if (count($v))
        {
            foreach ($v as $vv)
                print "remove $vv from $k\n";
        }
    }
}
echo "--------------ADD-----------------\n";
if (count($add))
{
    foreach ($add as $k => $v)
    {
        if (count($v))
        {
            foreach ($v as $vv)
                print "add $vv to $k\n";
        }
    }
}
echo "\n";

